File: frontend/src/services/axelarService.ts
import {
  AxelarAssetTransfer,
  AxelarQueryAPI,
  Environment,
  EvmChain,
  SendTokenParams
} from "@axelar-network/axelarjs-sdk";
import { ethers } from "ethers";
import { ChainUtils } from "../utils/chainUtils";
import { logger } from "../utils/logger";

export interface TransferQuote {
  fee: string;
  estimatedTime: string;
  route: string[];
}

export interface AxelarConfig {
  environment: Environment;
  rpcUrl?: string;
}

class AxelarService {
  private assetTransfer: AxelarAssetTransfer | null = null;
  private queryAPI: AxelarQueryAPI | null = null;
  private environment: Environment;
  private isInitialized: boolean = false;

  constructor(config: AxelarConfig = { environment: Environment.MAINNET }) {
    this.environment = config.environment;
    
    try {
      // Initialize AxelarAssetTransfer with proper config
      this.assetTransfer = new AxelarAssetTransfer({
        environment: this.environment,
        auth: "metamask" // Use metamask for browser environment
      });

      // Initialize AxelarQueryAPI
      this.queryAPI = new AxelarQueryAPI({
        environment: this.environment,
        axelarRpcUrl: config.rpcUrl
      });

      this.isInitialized = true;
      logger.service('axelar', `Service initialized successfully for ${this.environment}`);
    } catch (error) {
      logger.warn(`Failed to initialize Axelar service for ${this.environment}:`, error);
      
      // Fallback to testnet if mainnet fails
      if (this.environment === Environment.MAINNET) {
        try {
          this.environment = Environment.TESTNET;
          this.assetTransfer = new AxelarAssetTransfer({
            environment: this.environment,
            auth: "metamask"
          });
          this.queryAPI = new AxelarQueryAPI({
            environment: this.environment
          });
          this.isInitialized = true;
          logger.service('axelar', 'Service initialized with testnet fallback');
        } catch (fallbackError) {
          logger.error("Failed to initialize Axelar service with testnet fallback:", fallbackError);
          this.isInitialized = false;
        }
      } else {
        this.isInitialized = false;
      }
    }
  }

  /**
   * Check if the service is properly initialized
   */
  isReady(): boolean {
    return this.isInitialized && this.assetTransfer !== null && this.queryAPI !== null;
  }

  /**
   * Get the current environment
   */
  getEnvironment(): Environment {
    return this.environment;
  }

  /**
   * Convert chain name to Axelar format using ChainUtils
   */
  getAxelarChainName(chainNameOrId: string | number): string | null {
    try {
      let chainName: string;
      
      if (typeof chainNameOrId === 'number') {
        const config = ChainUtils.getChainConfig(chainNameOrId);
        chainName = config?.name || '';
      } else {
        chainName = chainNameOrId;
      }

      // Normalize the chain name and get Axelar mapping
      const normalizedName = ChainUtils.normalizeChainName(chainName);
      
      // Map common chain names to Axelar chain names
      const axelarMapping: { [key: string]: string } = {
        'Ethereum': 'ethereum',
        'Polygon': 'polygon',
        'Avalanche': 'avalanche',
        'Arbitrum': 'arbitrum',
        'optimism': 'optimism',
        'binance': 'binance',
        'base': 'base'
      };

      return axelarMapping[normalizedName] || null;
    } catch (error) {
      logger.error('Error getting Axelar chain name:', error);
      return null;
    }
  }

  /**
   * Check if a chain is supported by Axelar
   */
  isChainSupported(chainIdentifier: string | number): boolean {
    if (!this.isReady()) {
      return false;
    }

    try {
      const axelarName = this.getAxelarChainName(chainIdentifier);
      return axelarName !== null;
    } catch (error) {
      logger.error('Error checking chain support:', error);
      return false;
    }
  }

  /**
   * Get list of supported chains
   */
  getSupportedChains(): string[] {
    if (!this.isReady()) {
      return [];
    }

    try {
      const supportedChainIds = ChainUtils.getSupportedChainIds();
      return supportedChainIds
        .map(chainId => {
          const config = ChainUtils.getChainConfig(chainId);
          return config?.name;
        })
        .filter((name) => name !== undefined && this.getAxelarChainName(name) !== null) as string[];
    } catch (error) {
      logger.error('Error getting supported chains:', error);
      return [];
    }
  }

  /**
   * Get transfer quote for cross-chain transaction
   */
  async getTransferQuote(
    fromChain: string,
    toChain: string,
    asset: string,
    amount: string
  ): Promise<TransferQuote | null> {
    if (!this.isReady() || !this.queryAPI) {
      throw new Error('Axelar service not initialized');
    }

    try {
      const fromAxelarChain = this.getAxelarChainName(fromChain);
      const toAxelarChain = this.getAxelarChainName(toChain);

      if (!fromAxelarChain || !toAxelarChain) {
        throw new Error(`Unsupported chain: ${fromChain} -> ${toChain}`);
      }

      // Convert amount to number for the API
      const amountNumber = parseFloat(amount);
      if (isNaN(amountNumber)) {
        throw new Error('Invalid amount');
      }

      // Use the correct API method based on SDK structure
      const feeResponse = await Promise.race([
        this.queryAPI.getTransferFee(
          fromAxelarChain,
          toAxelarChain,
          asset,
          amountNumber
        ),
        new Promise((_, reject) => 
          setTimeout(() => reject(new Error('Request timeout')), 10000)
        )
      ]) as any;

      if (!feeResponse) {
        throw new Error('Failed to get transfer fee');
      }

      return {
        fee: feeResponse.fee?.amount || '0',
        estimatedTime: '10-20 minutes', // Default estimate
        route: [fromChain, toChain]
      };
    } catch (error) {
      logger.error('Error getting transfer quote:', error);
      throw error;
    }
  }

  /**
   * Get deposit address for cross-chain transfer
   */
  async getDepositAddress(
    fromChain: string,
    toChain: string,
    toAddress: string,
    asset: string
  ): Promise<string> {
    if (!this.isReady() || !this.assetTransfer) {
      throw new Error('Axelar service not initialized');
    }

    try {
      const fromAxelarChain = this.getAxelarChainName(fromChain);
      const toAxelarChain = this.getAxelarChainName(toChain);

      if (!fromAxelarChain || !toAxelarChain) {
        throw new Error(`Unsupported chain: ${fromChain} -> ${toChain}`);
      }

      // Use the correct method signature based on SDK
      const depositAddress = await this.assetTransfer.getDepositAddress({
        fromChain: fromAxelarChain,
        toChain: toAxelarChain,
        destinationAddress: toAddress,
        asset: asset
      });

      return depositAddress;
    } catch (error) {
      logger.error('Error getting deposit address:', error);
      throw error;
    }
  }

  /**
   * Execute cross-chain transfer
   */
  async executeTransfer(
    fromChain: string,
    toChain: string,
    toAddress: string,
    asset: string,
    amount: string,
    provider?: ethers.BrowserProvider
  ): Promise<string> {
    if (!this.isReady() || !this.assetTransfer) {
      throw new Error('Axelar service not initialized');
    }

    if (!provider) {
      throw new Error('Web3 provider required for transfer execution');
    }

    try {
      const fromAxelarChain = this.getAxelarChainName(fromChain);
      const toAxelarChain = this.getAxelarChainName(toChain);

      if (!fromAxelarChain || !toAxelarChain) {
        throw new Error(`Unsupported chain: ${fromChain} -> ${toChain}`);
      }

      const signer = await provider.getSigner();

      // Prepare the send token parameters according to the SDK interface
      const sendTokenParams: SendTokenParams = {
        fromChain: fromAxelarChain,
        toChain: toAxelarChain,
        destinationAddress: toAddress,
        asset: {
          symbol: asset
        },
        amountInAtomicUnits: amount,
        options: {
          evmOptions: {
            signer: signer as any,
            provider: provider as any
          }
        }
      };

      // Execute the transfer using the asset transfer service
      const txResponse = await this.assetTransfer.sendToken(sendTokenParams);

      // Extract transaction hash from the response
      if ('hash' in txResponse) {
        return txResponse.hash;
      } else if ('transactionHash' in txResponse) {
        return (txResponse as any).transactionHash;
      } else {
        throw new Error('Unable to extract transaction hash from response');
      }
    } catch (error) {
      logger.error('Error executing transfer:', error);
      throw error;
    }
  }

  /**
   * Get gas fee estimate for cross-chain transaction
   */
  async getGasFeeEstimate(
    fromChain: string,
    toChain: string,
    gasLimit: string = "100000"
  ): Promise<string> {
    if (!this.isReady() || !this.queryAPI) {
      throw new Error('Axelar service not initialized');
    }

    try {
      const fromAxelarChain = this.getAxelarChainName(fromChain);
      const toAxelarChain = this.getAxelarChainName(toChain);

      if (!fromAxelarChain || !toAxelarChain) {
        throw new Error(`Unsupported chain: ${fromChain} -> ${toChain}`);
      }

      // Use the estimateGasFee method
      const gasFee = await this.queryAPI.estimateGasFee(
        fromAxelarChain as EvmChain,
        toAxelarChain as EvmChain,
        gasLimit
      );

      return typeof gasFee === 'string' ? gasFee : gasFee.baseFee;
    } catch (error) {
      logger.error('Error getting gas fee estimate:', error);
      throw error;
    }
  }

  /**
   * Check if chains are active
   */
  async areChainsActive(chains: string[]): Promise<boolean> {
    if (!this.isReady() || !this.queryAPI) {
      return false;
    }

    try {
      const axelarChains = chains
        .map(chain => this.getAxelarChainName(chain))
        .filter((name): name is string => name !== null);

      if (axelarChains.length !== chains.length) {
        return false; // Some chains are not supported
      }

      // Check if all chains are active
      const activeChecks = await Promise.all(
        axelarChains.map(chain => this.queryAPI!.isChainActive(chain))
      );

      return activeChecks.every(isActive => isActive);
    } catch (error) {
      logger.error('Error checking chain activity:', error);
      return false;
    }
  }
}

// Export singleton instance
export const axelarService = new AxelarService({
  environment: process.env.NODE_ENV === 'production' ? Environment.MAINNET : Environment.TESTNET
});

export default AxelarService;

File: backend/app/services/axelar_service.py
"""
Axelar service for cross-chain operations - ENHANCED VERSION.
Real Axelar SDK integration with proper error handling and configuration management.
"""
import logging
import httpx
import asyncio
import time
from typing import Dict, Any, Optional, List, Tuple
from decimal import Decimal
import os
import json
from eth_abi import encode as abi_encode
from eth_utils import function_signature_to_4byte_selector, to_checksum_address
from ..core.config_manager import config_manager
from ..core.errors import ProtocolError, ProtocolAPIError, NetworkError, ValidationError

logger = logging.getLogger(__name__)

class AxelarService:
    """Enhanced Axelar service with real SDK integration and proper error handling."""

    def __init__(self):
        """Initialize Axelar service."""
        self.config = None
        self.session = None
        # Rate limiting and circuit breaker
        self._rate_limit_window = 1.0  # seconds
        self._max_requests_per_window = 5  # Conservative for Axelar API
        self._failure_threshold = 3
        self._cooldown_seconds = 60
        self._api_state = {}  # Track API health per endpoint
        # Response cache for fee estimates
        self._fee_cache = {}
        self._fee_cache_ttl = 300  # 5 minutes
        
        # Initialize chain mappings immediately
        self.chain_mappings = {
            1: "ethereum",
            10: "optimism", 
            56: "binance",
            137: "polygon",
            43114: "avalanche",
            42161: "arbitrum",
            8453: "base",
            59144: "linea",
            534352: "scroll"
        }

    async def initialize(self):
        """Initialize service with configuration manager."""
        self.config = await config_manager.get_protocol("axelar")
        if not self.config:
            raise ProtocolError(
                message="Axelar protocol configuration not found",
                protocol="axelar",
                user_message="Axelar bridge is not properly configured"
            )

        # Initialize HTTP session
        timeout = httpx.Timeout(30.0, connect=5.0)
        headers = {
            "User-Agent": "SNEL/1.0"
        }
        self.session = httpx.AsyncClient(timeout=timeout, headers=headers)

        # Remove hardcoded token addresses - use ConfigurationManager instead
        logger.info("Axelar service initialized - using ConfigurationManager for token addresses")

    async def _apply_rate_limit(self, endpoint: str):
        """Apply rate limiting per API endpoint."""
        state = self._api_state.setdefault(endpoint, {"timestamps": [], "failures": 0, "circuit_open": False, "circuit_opened_at": 0})
        now = time.time()
        
        # Remove old timestamps
        state["timestamps"] = [t for t in state["timestamps"] if now - t < self._rate_limit_window]
        
        if len(state["timestamps"]) >= self._max_requests_per_window:
            sleep_time = self._rate_limit_window - (now - min(state["timestamps"]))
            await asyncio.sleep(sleep_time)
        
        state["timestamps"].append(time.time())

    def _record_failure(self, endpoint: str):
        """Record API failure and open circuit if threshold exceeded."""
        state = self._api_state.setdefault(endpoint, {"timestamps": [], "failures": 0, "circuit_open": False, "circuit_opened_at": 0})
        state["failures"] += 1
        
        if state["failures"] >= self._failure_threshold:
            state["circuit_open"] = True
            state["circuit_opened_at"] = time.time()
            logger.warning(f"Circuit breaker opened for Axelar endpoint: {endpoint}")

    def _reset_circuit(self, endpoint: str):
        """Reset circuit breaker after cooldown period."""
        state = self._api_state.get(endpoint)
        if state and state["circuit_open"]:
            if time.time() - state["circuit_opened_at"] >= self._cooldown_seconds:
                state["circuit_open"] = False
                state["failures"] = 0
                logger.info(f"Circuit breaker reset for Axelar endpoint: {endpoint}")

    async def _api_call(self, endpoint: str, method: str = "GET", **kwargs) -> Dict[str, Any]:
        """Make API call with rate limiting and circuit breaker."""
        self._reset_circuit(endpoint)
        state = self._api_state.get(endpoint, {})
        
        if state.get("circuit_open"):
            raise ProtocolAPIError(
                protocol="axelar",
                endpoint=endpoint,
                status_code=0,
                user_message="Axelar API temporarily unavailable"
            )
        
        await self._apply_rate_limit(endpoint)
        
        try:
            if method.upper() == "GET":
                response = await self.session.get(endpoint, **kwargs)
            else:
                response = await self.session.post(endpoint, **kwargs)
            
            response.raise_for_status()
            
            # Reset failures on success
            if state:
                state["failures"] = 0
            
            return response.json()
            
        except Exception as e:
            self._record_failure(endpoint)
            raise e

    async def get_axelar_chain_name(self, chain_id: int) -> Optional[str]:
        """Get Axelar chain name from chain ID using ConfigurationManager."""
        if not self.config:
            await self.initialize()

        # Use our chain mapping for Axelar-specific names
        # The chain_mappings dict is the authoritative source for Axelar support
        return self.chain_mappings.get(chain_id)

    def is_chain_supported(self, chain_id: int) -> bool:
        """Check if chain is supported by Axelar."""
        return chain_id in self.chain_mappings

    async def is_token_supported(self, token_symbol: str, chain_id: int) -> bool:
        """Check if token is supported on the given chain via Axelar."""
        chain_name = await self.get_axelar_chain_name(chain_id)
        if not chain_name:
            return False

        # Check if token exists in our configuration
        token_config = await config_manager.get_token_by_symbol(token_symbol)
        if not token_config:
            return False

        # Check if token has address on this chain
        return chain_id in token_config.addresses

    async def get_cross_chain_quote(
        self,
        from_token: str,
        to_token: str,
        amount: Decimal,
        from_chain_id: int,
        to_chain_id: int,
        wallet_address: str
    ) -> Dict[str, Any]:
        """
        Get quote for cross-chain transfer using Axelar.

        Args:
            from_token: Source token symbol
            to_token: Destination token symbol
            amount: Amount to transfer
            from_chain_id: Source chain ID
            to_chain_id: Destination chain ID
            wallet_address: User wallet address

        Returns:
            Quote information or error
        """
        try:
            if not self.session:
                await self.initialize()

            # Enhanced validation using new validation method
            from_chain = await self.get_axelar_chain_name(from_chain_id)
            to_chain = await self.get_axelar_chain_name(to_chain_id)

            if not from_chain or not to_chain:
                raise ProtocolError(
                    message=f"Unsupported chain(s): {from_chain_id} -> {to_chain_id}",
                    protocol="axelar",
                    user_message="One or both chains are not supported by Axelar"
                )

            # Comprehensive validation
            is_valid, validation_message = await self._validate_cross_chain_transfer(
                from_chain, to_chain, from_token, wallet_address
            )
            
            if not is_valid:
                raise ValidationError(
                    message=validation_message,
                    field="cross_chain_transfer",
                    value=f"{from_chain} -> {to_chain}"
                )

            # For cross-chain transfers, tokens should typically be the same
            if from_token.upper() != to_token.upper():
                raise ValidationError(
                    message="Cross-chain swaps between different tokens not supported",
                    field="token_pair",
                    value=f"{from_token} -> {to_token}"
                )

            # Get transfer fee estimate
            fee_estimate = await self._get_transfer_fee(from_chain, to_chain, from_token, float(amount))

            # Calculate estimated time
            estimated_time = self._get_estimated_time(from_chain, to_chain)

            # Get Axelar Gateway address for the source chain
            gateway_address = await self._get_gateway_address(from_chain_id)
            if not gateway_address:
                raise ProtocolError(
                    message=f"Axelar Gateway not configured for chain {from_chain_id}",
                    protocol="axelar",
                    user_message="Axelar bridge not available on this network"
                )

            # Get token address for the source chain
            token_address = await self._get_token_address(from_token, from_chain_id)
            if not token_address:
                raise ProtocolError(
                    message=f"Token {from_token} not found on chain {from_chain_id}",
                    protocol="axelar",
                    user_message=f"Token {from_token} not supported on this network"
                )

            return {
                "success": True,
                "protocol": "axelar",
                "type": "cross_chain_transfer",
                "from_chain": from_chain,
                "to_chain": to_chain,
                "from_token": from_token,
                "to_token": to_token,
                "amount": str(amount),
                "estimated_fee": fee_estimate,
                "estimated_time": estimated_time,
                "gateway_address": gateway_address,
                "token_address": token_address,
                "recipient_address": wallet_address,
                "steps": [
                    {
                        "type": "approve",
                        "description": f"Approve {amount} {from_token} for Axelar Gateway",
                        "chain": from_chain,
                        "to": token_address,
                        "value": "0",
                        "data": await self._build_approve_data(gateway_address, amount, from_token, from_chain_id)
                    },
                    {
                        "type": "send_token",
                        "description": f"Send {amount} {from_token} to {to_chain} via Axelar",
                        "chain": from_chain,
                        "to": gateway_address,
                        "value": "0",
                        "data": await self._build_send_token_data(to_chain, wallet_address, from_token, amount)
                    }
                ]
            }

        except Exception as e:
            logger.exception(f"Error getting Axelar quote: {e}")
            return {
                "error": "Failed to get cross-chain quote",
                "technical_details": str(e)
            }

    async def get_same_chain_quote(
        self,
        from_token: str,
        to_token: str,
        amount: Decimal,
        chain_id: int,
        wallet_address: str
    ) -> Dict[str, Any]:
        """
        Get quote for same-chain swap.
        Note: Axelar is primarily for cross-chain, so this will return an error
        suggesting to use a different protocol for same-chain swaps.
        """
        chain_name = await self.get_axelar_chain_name(chain_id)

        return {
            "error": "Axelar is designed for cross-chain transfers. For same-chain swaps, please use a different protocol.",
            "technical_details": f"Same-chain swap requested on {chain_name} ({chain_id})",
            "suggestion": "Try using a DEX aggregator like 1inch or 0x for same-chain swaps."
        }

    async def _get_transfer_fee(self, from_chain: str, to_chain: str, token: str, amount: float) -> str:
        """Get transfer fee estimate using real Axelar API endpoints."""
        try:
            if not self.session:
                await self.initialize()

            # Cache key for fee estimates
            cache_key = f"{from_chain}:{to_chain}:{token}:{int(amount)}"
            cached = self._fee_cache.get(cache_key)
            now = int(time.time())
            
            if cached and (now - cached.get("ts", 0) <= self._fee_cache_ttl):
                return cached["fee"]

            # Use real Axelar API endpoint for gas estimation
            api_endpoint = self.config.api_endpoints.get("mainnet", "https://api.axelar.dev")
            
            try:
                # Real Axelar API endpoint for gas estimation
                endpoint = f"{api_endpoint}/gas-price/{from_chain.lower()}/{to_chain.lower()}/{token.upper()}"
                
                data = await self._api_call(endpoint, method="GET")
                
                # Extract gas price and estimate fee
                gas_price = data.get("gas_price", "0.01")  # Default fallback
                
                # Cache the result
                self._fee_cache[cache_key] = {"ts": now, "fee": str(gas_price)}
                
                return str(gas_price)
                
            except Exception as api_error:
                logger.warning(f"Axelar API call failed: {api_error}")
                
                # Intelligent fallback based on chain characteristics
                base_fee = self._estimate_base_fee(from_chain, to_chain, token, amount)
                
                # Cache fallback result with shorter TTL
                self._fee_cache[cache_key] = {"ts": now, "fee": base_fee}
                
                return base_fee

        except Exception as e:
            logger.error(f"Error getting transfer fee: {e}")
            # Final fallback
            return self._estimate_base_fee(from_chain, to_chain, token, amount)

    def _estimate_base_fee(self, from_chain: str, to_chain: str, token: str, amount: float) -> str:
        """Intelligent fee estimation based on chain characteristics."""
        # Base fees by chain (in USD)
        chain_base_fees = {
            "ethereum": 0.05,  # Higher due to gas costs
            "polygon": 0.01,   # Lower gas costs
            "arbitrum": 0.02,  # L2 efficiency
            "optimism": 0.02,  # L2 efficiency
            "avalanche": 0.02, # Fast finality
            "base": 0.01,      # L2 efficiency
            "binance": 0.01,   # Low fees
            "linea": 0.02,     # New L2
            "scroll": 0.02     # New L2
        }
        
        from_fee = chain_base_fees.get(from_chain.lower(), 0.03)
        to_fee = chain_base_fees.get(to_chain.lower(), 0.03)
        
        # Cross-chain premium
        cross_chain_premium = 0.02
        
        # Amount-based scaling (higher amounts = higher absolute fees)
        amount_factor = min(amount * 0.001, 0.05)  # Cap at 5% of amount
        
        total_fee = from_fee + to_fee + cross_chain_premium + amount_factor
        
        return f"{total_fee:.4f}"

    async def _validate_cross_chain_transfer(self, from_chain: str, to_chain: str, token: str, recipient: str) -> Tuple[bool, str]:
        """Validate cross-chain transfer parameters."""
        # Validate recipient address format
        if not recipient or len(recipient) != 42 or not recipient.startswith("0x"):
            return False, f"Invalid recipient address format: {recipient}"
        
        # Validate chains are different
        if from_chain.lower() == to_chain.lower():
            return False, "Source and destination chains must be different for cross-chain transfers"
        
        # Validate both chains are supported
        from_chain_id = None
        to_chain_id = None
        
        for chain_id, chain_name in self.chain_mappings.items():
            if chain_name.lower() == from_chain.lower():
                from_chain_id = chain_id
            if chain_name.lower() == to_chain.lower():
                to_chain_id = chain_id
        
        if not from_chain_id:
            return False, f"Source chain '{from_chain}' not supported by Axelar"
        
        if not to_chain_id:
            return False, f"Destination chain '{to_chain}' not supported by Axelar"
        
        # Validate token is supported on both chains
        token_config = await config_manager.get_token_by_symbol(token)
        if not token_config:
            return False, f"Token '{token}' not found in configuration"
        
        if from_chain_id not in token_config.addresses:
            return False, f"Token '{token}' not supported on {from_chain}"
        
        if to_chain_id not in token_config.addresses:
            return False, f"Token '{token}' not supported on {to_chain}"
        
        return True, "Validation successful"

    def _get_estimated_time(self, from_chain: str, to_chain: str) -> str:
        """Get estimated transfer time based on chains."""
        # Ethereum transfers typically take longer due to finality requirements
        if from_chain == "Ethereum" or to_chain == "Ethereum":
            return "15-20 minutes"
        else:
            return "5-10 minutes"

    def get_supported_chains(self) -> Dict[int, str]:
        """Get all supported chains."""
        return self.chain_mappings.copy()

    async def get_supported_tokens_for_chain(self, chain_id: int) -> List[str]:
        """Get list of tokens supported on a specific chain."""
        if not await self.is_chain_supported(chain_id):
            return []

        supported = []
        all_tokens = await config_manager.get_all_tokens()
        for token_id, token_config in all_tokens.items():
            if chain_id in token_config.addresses:
                supported.append(token_config.symbol)
        return supported

    async def close(self):
        """Clean up resources."""
        if self.session:
            await self.session.aclose()

    async def _get_gateway_address(self, chain_id: int) -> Optional[str]:
        """Get Axelar Gateway contract address from ConfigurationManager."""
        if not self.config:
            await self.initialize()
        
        # Get gateway address from configuration
        contracts = self.config.contract_addresses.get(chain_id, {})
        return contracts.get("gateway")

    async def _get_token_address(self, token_symbol: str, chain_id: int) -> Optional[str]:
        """Get token contract address from ConfigurationManager."""
        token_config = await config_manager.get_token_by_symbol(token_symbol)
        if not token_config:
            return None
        
        return token_config.addresses.get(chain_id)

    async def _build_approve_data(self, spender: str, amount: Decimal, token_symbol: str, chain_id: int) -> str:
        """Build approve transaction data for ERC20 token using ConfigurationManager."""
        try:
            # Get token configuration for accurate decimals
            token_config = await config_manager.get_token_by_symbol(token_symbol)
            if not token_config:
                raise ValueError(f"Token {token_symbol} not found in configuration")
            
            decimals = token_config.decimals
            amount_wei = int(amount * (Decimal(10) ** decimals))

            # ERC20 approve function: approve(address spender, uint256 amount)
            function_selector = function_signature_to_4byte_selector("approve(address,uint256)").hex()

            # Encode parameters: spender address and amount
            encoded_params = abi_encode(['address', 'uint256'], [to_checksum_address(spender), amount_wei])

            return f"0x{function_selector}{encoded_params.hex()}"

        except Exception as e:
            logger.error(f"Error building approve data: {e}")
            raise ProtocolError(
                message=f"Failed to build approve transaction: {e}",
                protocol="axelar",
                user_message="Unable to prepare token approval transaction"
            )

    async def _build_send_token_data(self, dest_chain: str, recipient: str, token_symbol: str, amount: Decimal) -> str:
        """Build sendToken transaction data for Axelar Gateway using ConfigurationManager."""
        try:
            # Get token configuration for accurate decimals
            token_config = await config_manager.get_token_by_symbol(token_symbol)
            if not token_config:
                raise ValueError(f"Token {token_symbol} not found in configuration")
            
            decimals = token_config.decimals
            amount_wei = int(amount * (Decimal(10) ** decimals))

            # Axelar Gateway sendToken function: sendToken(string destinationChain, string destinationAddress, string symbol, uint256 amount)
            function_selector = function_signature_to_4byte_selector("sendToken(string,string,string,uint256)").hex()

            # Encode parameters with proper formatting
            encoded_params = abi_encode(
                ['string', 'string', 'string', 'uint256'],
                [dest_chain.lower(), to_checksum_address(recipient), token_symbol.upper(), amount_wei]
            )

            return f"0x{function_selector}{encoded_params.hex()}"

        except Exception as e:
            logger.error(f"Error building send token data: {e}")
            raise ProtocolError(
                message=f"Failed to build send token transaction: {e}",
                protocol="axelar",
                user_message="Unable to prepare cross-chain transfer transaction"
            )

# Global instance
axelar_service = AxelarService()
